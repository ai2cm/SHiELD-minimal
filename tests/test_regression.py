import hashlib
import os
import subprocess
import typing

import fv3config
import pytest

from pathlib import Path


CONFIG_DIR = Path(__file__).parent
EXECUTABLE = Path("/SHiELD/SHiELD_build/Build/bin/SHiELD_nh.prod.64bit.gnu.x")


def get_config(filename):
    config_filename = CONFIG_DIR / filename
    with open(config_filename, "r") as f:
        return fv3config.load(f)


def checksum_file(path: Path) -> str:
    sum = hashlib.md5()
    BUFFER_SIZE = 1024 * 1024
    with open(path, "rb") as f:
        while True:
            buf = f.read(BUFFER_SIZE)
            if not buf:
                break
            sum.update(buf)
    return sum.hexdigest()


def get_rundir_netcdfs(rundir: Path) -> typing.List[Path]:
    restart_directory = rundir / "RESTART"
    diagnostics_files = sorted(rundir.glob("*.nc"))
    restart_files = sorted(restart_directory.glob("*.nc"))
    return diagnostics_files + restart_files    
        

def checksum_rundir_to_file(rundir: Path, file: Path):
    """checksum rundir storing output in file"""
    rundir_netcdfs = get_rundir_netcdfs(rundir)
    for path in rundir_netcdfs:
        print(path, checksum_file(path), file=file)


def _run_simulation(config: dict, rundir: Path, command: typing.List[str]):
    fv3config.write_run_directory(config, rundir)
    n_processes = fv3config.config.get_n_processes(config)
    command = ["mpirun", "-n", str(n_processes)] + command
    completed_process = subprocess.run(command, cwd=rundir, capture_output=True)
    if completed_process.returncode != 0:
        print("Tail of Stderr:")
        print(completed_process.stderr[-2000:].decode())
        print("Tail of Stdout:")
        print(completed_process.stdout[-2000:].decode())
        pytest.fail()
    return completed_process
    
        
def run_fortran_executable(config: dict, rundir: Path):
    command = [EXECUTABLE.absolute().as_posix()]
    return _run_simulation(config, rundir, command)    


def test_regression_fortran(tmp_path: Path, regtest):
    config = get_config("default.yml")
    rundir_fortran = tmp_path / "rundir-fortran"

    run_fortran_executable(config, rundir_fortran)
    checksum_rundir_to_file(rundir_fortran, file=regtest)
