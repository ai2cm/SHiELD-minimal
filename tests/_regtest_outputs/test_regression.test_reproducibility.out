<pytest_tempdir>/test_reproducibility0/rundir/atmos_dt_atmos.tile1.nc dd90e954cfeccd43ac90909282d806c7
<pytest_tempdir>/test_reproducibility0/rundir/atmos_dt_atmos.tile2.nc ee2fdbafa737be8db4fb3be19bd9926f
<pytest_tempdir>/test_reproducibility0/rundir/atmos_dt_atmos.tile3.nc ec1b20dd1c192626fe320f69aa40ab47
<pytest_tempdir>/test_reproducibility0/rundir/atmos_dt_atmos.tile4.nc 517c38515bf232775ddff8770e684562
<pytest_tempdir>/test_reproducibility0/rundir/atmos_dt_atmos.tile5.nc 7ecf63f52bc0b9c517954759b3209989
<pytest_tempdir>/test_reproducibility0/rundir/atmos_dt_atmos.tile6.nc 4fa14ab42fd9cde6771a5547eb54b54a
<pytest_tempdir>/test_reproducibility0/rundir/atmos_static.tile1.nc 0dc74e89e2bd6ee5e91ac55ed96f30bd
<pytest_tempdir>/test_reproducibility0/rundir/atmos_static.tile2.nc c8e66f31407e4b360f40e8f2d6348cd7
<pytest_tempdir>/test_reproducibility0/rundir/atmos_static.tile3.nc 8efd8a973ac687ed854b495f3a279bad
<pytest_tempdir>/test_reproducibility0/rundir/atmos_static.tile4.nc 85a260fdcdf3748f47b6a5c530e145e2
<pytest_tempdir>/test_reproducibility0/rundir/atmos_static.tile5.nc 822ba7caefdc00317d4058b9e8dd4fa5
<pytest_tempdir>/test_reproducibility0/rundir/atmos_static.tile6.nc 664dfed0eaafb54f911b4acada7c9525
<pytest_tempdir>/test_reproducibility0/rundir/sfc_dt_atmos.tile1.nc edb2b5f8a6aef24bcf65c062b42ef435
<pytest_tempdir>/test_reproducibility0/rundir/sfc_dt_atmos.tile2.nc b6d80c2a8ba03fd3a060d117341ccee7
<pytest_tempdir>/test_reproducibility0/rundir/sfc_dt_atmos.tile3.nc bd658b5209597dc8115918c56b6f8620
<pytest_tempdir>/test_reproducibility0/rundir/sfc_dt_atmos.tile4.nc 7815455ced966fba41e39c8c2b18f065
<pytest_tempdir>/test_reproducibility0/rundir/sfc_dt_atmos.tile5.nc aa8371a03e194ba0dd3faa8584b4c9d8
<pytest_tempdir>/test_reproducibility0/rundir/sfc_dt_atmos.tile6.nc 9a37d413be3b7e4aea57c985634f6103
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_core.res.nc 18f136825b9be2c9daad27aff6e25375
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_core.res.tile1.nc 4e7e42ed0d462148537880754d6f37c6
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_core.res.tile2.nc dec764c57a331d65f07d20bcdc9da22d
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_core.res.tile3.nc 9ccee287b43ca5c56d53b8d563e583b3
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_core.res.tile4.nc 40d45140cdb3b4e05f8cadef40ae54e7
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_core.res.tile5.nc 5194c8107c2ac6b5c2d230e2fb16e91f
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_core.res.tile6.nc 96cba570a2866b3d419e5446b76ac2af
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_srf_wnd.res.tile1.nc f789d74a6abb0706771e4d667dbe95b9
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_srf_wnd.res.tile2.nc 9fc7b625f8c50b78c44322a8168a10db
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_srf_wnd.res.tile3.nc 91b80e8bc3bf631b77512a665d7293a4
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_srf_wnd.res.tile4.nc 4cf8391459c43abd9c340df816ea2cbf
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_srf_wnd.res.tile5.nc f75991a82cc11257a95a4bcdb6b233ee
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_srf_wnd.res.tile6.nc 7a613a6c69bca69e23d8e81b8b145288
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_tracer.res.tile1.nc 6f4b0b1bc4175be9a48ed258fea81e2a
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_tracer.res.tile2.nc 4499cc565e963556102ac3f889f27613
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_tracer.res.tile3.nc f6292eef07f466687059fd340c9f6aff
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_tracer.res.tile4.nc 0d9902fb0ef5ff27baa68c8fa555d39b
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_tracer.res.tile5.nc 78c3f6395a4a293926f23a208550faff
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/fv_tracer.res.tile6.nc a79598edadfd9ca8e9124d794833a3ee
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/phy_data.tile1.nc c907b002740b7090e7752fcfbc0365c8
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/phy_data.tile2.nc ebf06b07b304eeef39e5be80be91beaa
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/phy_data.tile3.nc fad4c687044b81103a2efaf5e7f17c76
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/phy_data.tile4.nc 14f25ea7fbd979ff91108a15096341c2
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/phy_data.tile5.nc b52a13593daa5322a4f726ef9e8a344c
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/phy_data.tile6.nc 8c1ddf1db8cb4ba7df8e1b3377db8721
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/sfc_data.tile1.nc 58682e8ef0e23ff37683817c933741fe
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/sfc_data.tile2.nc cecbb0f259ad75acd350fdf4433f8ef6
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/sfc_data.tile3.nc 8a979962dc8d38896a1811cc7767fe0b
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/sfc_data.tile4.nc defa5dbcaf6319d71bc32da604d1bcb3
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/sfc_data.tile5.nc 6af18bb5b25e8a9a7b8ae7bd6aa06025
<pytest_tempdir>/test_reproducibility0/rundir/RESTART/sfc_data.tile6.nc a100eefddf9e09fd70ee495749459575
